### PR TITLE
Find: Improve precision of benchmarking

### PR DESCRIPTION
**Pull request by atamazov**
This PR simpifies the code and improves the quality of benchmarking during Find calls. This opens the opportunity to improve performance by an average of 2.5% and for free, i.e. without any new solvers, etc. We simply need to retune find-db for that (which we should do periodically anyway). Details:

- Library:
  - [Improvement] Find API calls: Run invoker 5 times and return average.
  - [Improvement] Find API calls: No special "Evaluate" type of runs (dedicated for benchmarking), make a dry-run of invoker instead.
    - Why: "Evaluate" runs were invented long ago and used in GEMM conv solvers only, to warm-up rocBlas. In fact, this feature adds a dry-run of each rocBlas primitive prior actual benchmarking. Dry-running invokers has the same effect while allowing us to:
      - Have generic warm-up that works with all solvers.
      - Clean up GEMM stuff, including:
        - `InvokeType::Evaluate`
        - `MIOPEN_CONV_PRECISE_ROCBLAS_TIMING`
        - Invocation paths in GEMM conv solvers dedicated for benchmarking
        - The `CallGemmTimeMeasure()` wrapper
  - [Fix] Resolved issue #2967.
- Driver
  - [Improvement] Run API calls: Throw away the 1st wall time to improve precision of Wall Time measurements.

## :cyclone: Routine performance/correctness testing

- Preconditions: Ubuntu 20.04.5 LTS, BUILD_DEV=On, Release build, Navi21/36, ROCm 6.0.2 docker
- Versions of MIOpen under test: `develop` a826e210f with 3fe519072 applied vs branch adef229f6.
- Tested configs:
  - 93 known FP32 convolution configs
  - 93 known FP16 convolution configs
- Modes:
  - (1) Default Find mode (DYNAMIC_HYBRID) with default system find-db
  - (2) Immediate mode (with user find-db possibly updated at step (1))
- 🟢 No performance or correctness regressions.

## :cyclone: Feature performance/correctness testing

The idea of the test is to find out how much performance can we get from better benchmarking.

First, a set of convolutions is run with baseline. Convolutions are run in NORMAL Find mode, thus finding out the best solvers using current benchmarking machinery. This simulates re-tuning of find-db. Then, with regenerated find-db, metrics are taken in the default Find mode and in Immediate mode.

Second, the same is repeated with PR.

Third, we compare metrics and see the difference.

- Preconditions: Ubuntu 20.04.5 LTS, BUILD_DEV=On, Release build, Navi21/36, ROCm 6.0.2 docker
- Versions of MIOpen under test: `develop` a826e210f with 3fe519072 applied vs branch adef229f6.
- Tested configs:
  - 93 known FP32 convolution configs
  - 93 known FP16 convolution configs
- Modes:
  - (1) NORMAL Find mode (DYNAMIC_HYBRID) with default system find-db
  - (2) Default Find mode (DYNAMIC_HYBRID) with user find-db regenerated step (1)
  - (3) Immediate mode with user find-db regenerated step (1)
- 🟢 Performance gain ~2.5% (on average). No correctness regressions.

&nbsp; | &nbsp; | &nbsp; | GPU time | &nbsp; | &nbsp; | All directions | &nbsp; | &nbsp; | &nbsp; | &nbsp; 
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
&nbsp; | Mode | Find-db | Fwd | Bwd | WrW | Avg GPU time | Sum GPU time | Avg Wall time | Sum Wall time | Sum AUX Wall time
FP32 | Norm Find | * | 0.2% | 5.9% | 3.4% | 3.2% | 3.9% | 2.0% | 1.3% | -11.2%
&nbsp; | Dflt Find | Regen | 0.6% | 3.2% | 4.4% | 2.7% | 4.1% | 1.1% | 1.3% | 2.3%
&nbsp; | Imm | Regen | -0.5% | 3.3% | 3.2% | 2.0% | 3.3% | 0.4% | 0.6% | 2.6%
FP16 | Norm Find | * | 4.6% | 4.9% | 1.3% | 3.6% | 2.1% | 2.0% | -0.1% | -11.3%
&nbsp; | Dflt Find | Regen | 3.5% | 3.2% | 1.0% | 2.5% | 2.1% | -0.5% | -0.7% | 31.7%
&nbsp; | Imm | Regen | 3.4% | 2.2% | 1.7% | 2.4% | 2.3% | -0.3% | -0.3% | 38.0%

----
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCm/MIOpen/labels/enhancement
- https://github.com/ROCm/MIOpen/labels/performance
- https://github.com/ROCm/MIOpen/labels/urgency_normal
- https://github.com/ROCm/MIOpen/labels/value_middle
- Proposed reviewers: @JehandadKhan @DrizztDoUrden @averinevg 
- Optional reviewers: @CAHEK7 @cderb @junliume @asroy @ce1adon @shurale-nkn

### CODE REVIEW

**Comment by atamazov** [comment]
@junliume @JehandadKhan @DrizztDoUrden @averinevg PR is ready for CI testing (please launch the job, thanks!) and reviewing.

File: src/conv/solver_finders.cpp [hunk 1]
```diff
@@ -257,8 +257,18 @@ static void EvaluateInvokers(Handle& handle,
...
+            constexpr int N_RUNS = 5;
+            using elapsed_t      = decltype(handle.GetKernelTime());
+            auto elapsed         = static_cast<elapsed_t>(0);
+            for(int i = 0; i < N_RUNS; ++i)
+            {
+                invoker(handle, invoke_ctx);
+                elapsed += handle.GetKernelTime();
+            }
+            elapsed /= static_cast<elapsed_t>(N_RUNS);
+
```
Comment (atamazov) [inline src/conv/solver_finders.cpp:271]: This is the core of this change. Other stuff is just removals of unused code.

**Comment by junliume** [comment]
CI has started, will watch closely :)

**Comment by atamazov** [comment]
@junliume PR is ready for another CI run.

**Review by junliume — APPROVED** [review]

**Comment by atamazov** [comment]
Ping to
- Proposed reviewers: @JehandadKhan @DrizztDoUrden @averinevg
- Optional reviewers: @CAHEK7 @cderb @junliume @asroy @ce1adon @shurale-nkn

File: src/solver/gemm.cpp [hunk 2]
```diff
@@ -40,8 +40,6 @@
 #include <boost/any.hpp>
 #include <boost/range/adaptors.hpp>
 
-MIOPEN_DECLARE_ENV_VAR_BOOL(MIOPEN_CONV_PRECISE_ROCBLAS_TIMING)
```
Comment (averinevg) [inline src/solver/gemm.cpp:43]: I have now looked in the library files for this environment variable, it is still there in some files. Let's remove it everywhere.

File: driver/conv_driver.hpp [hunk 3]
```diff
@@ -1786,9 +1793,7 @@ template <typename Tgpu, typename Tref>
 void ConvDriver<Tgpu, Tref>::PrintForwardTime(const float kernel_total_time,
                                               const float kernel_first_time) const
 {
-    float kernel_average_time = num_iterations > 1
-                                    ? (kernel_total_time - kernel_first_time) / (num_iterations - 1)
-                                    : kernel_first_time;
```
Comment (averinevg) [inline driver/conv_driver.hpp:1791]: This thing was copy-pasted throughout the entire `driver`, perhaps it’s worth replacing in other places too?

File: driver/conv_driver.hpp [hunk 4]
```diff
@@ -2149,13 +2154,19 @@ int ConvDriver<Tgpu, Tref>::RunForwardGpuFind(const bool is_transform)
 
     float alpha = static_cast<float>(1), beta = static_cast<float>(0);
 
-    float kernel_total_time = 0.0;
-    float kernel_first_time = 0.0;
+    float kernel_total_time = 0.f;
```
Comment (averinevg) [inline driver/conv_driver.hpp:2157]: This is not specific to this pull request, but why are we using a float data type for time intervals?

File: src/conv/solver_finders.cpp (hunk 1 above)
Comment (averinevg) [reply src/conv/solver_finders.cpp:271]: For cases where the execution time is relatively long (hundreds of milliseconds), the influence of external factors is negligible. Maybe in such cases one launch is enough?

**Review by junliume — APPROVED** [review]
@averinevg if the reviews are not critical (blocking), could you help to create a separate issue to track? Thanks!

**Comment by averinevg** [comment]
> @averinevg if the reviews are not critical (blocking), could you help to create a separate issue to track? Thanks!

@junliume This is not critical

File: driver/conv_driver.hpp (hunk 4 above)
Comment (atamazov) [reply driver/conv_driver.hpp:2157]: @averinevg Tradition. Any proposals?

File: driver/conv_driver.hpp (hunk 3 above)
Comment (atamazov) [reply driver/conv_driver.hpp:1791]: @averinevg Do you mean changes in driver modules intended for primitives other than convolution?

Comment (averinevg) [reply driver/conv_driver.hpp:1791]: @atamazov Yes.

Comment (atamazov) [reply driver/conv_driver.hpp:1791]: @averinevg Sure, this idea is good, but this is out of the scope of this PR. Invention of a good abstraction for this kind of measurements, suitable for all primitives (yeah, _how_ should we measure time _depends on the primitive_ in certain cases) is a substantial work.

**Comment by atamazov** [comment]
@averinevg 
> For cases where the execution time is relatively long (hundreds of milliseconds), the influence of external factors is negligible. Maybe in such cases one launch is enough?

Maybe, but this depends on too many factors and must be proven before use. But the idea is good.

File: src/conv/solver_finders.cpp (hunk 1 above)
Comment (atamazov) [reply src/conv/solver_finders.cpp:271]: @averinevg Maybe, but this depends on too many factors and must be proven before use. But the idea is good.

Comment (atamazov) [reply src/conv/solver_finders.cpp:271]: > For cases where the execution time is relatively long (hundreds of milliseconds), the influence of external factors is negligible. Maybe in such cases one launch is enough?

Resolved in #3103

**Pull request merged**